File: ios/Sources/HRCore/Sync/Outbox.swift
import Foundation
import GRDB

/// The write side of offline sync.
///
/// Implements docs/sync-protocol.md §4, mirroring `Outbox` on Android. Callers enqueue a mutation
/// and return immediately; the UI has already been updated from the local write that accompanied
/// it.
public actor Outbox {
    private let database: DatabaseQueue
    private let clock: @Sendable () -> Date

    public init(database: DatabaseQueue, clock: @escaping @Sendable () -> Date = { Date() }) {
        self.database = database
        self.clock = clock
    }

    /// Queues a mutation.
    ///
    /// The idempotency key is generated here, once. It must never be regenerated on retry — that
    /// is precisely what makes unlimited retries safe, and regenerating it would turn a lost
    /// response into a duplicate record.
    @discardableResult
    public func enqueue(
        aggregateType: String,
        aggregateId: String,
        httpMethod: String,
        path: String,
        payload: String
    ) async throws -> String {
        var entry = OutboxEntry(
            aggregateType: aggregateType,
            aggregateId: aggregateId,
            httpMethod: httpMethod,
            path: path,
            payload: payload,
            createdAt: clock()
        )
        try await database.write { db in
            try entry.insert(db)
        }
        return entry.id
    }

    /// The next entry to send for each aggregate.
    ///
    /// One per aggregate is what gives ordering within an aggregate and concurrency across them
    /// (docs/sync-protocol.md §4.2).
    public func nextBatch(limit: Int = 20) async throws -> [OutboxEntry] {
        let now = clock()
        return try await database.read { db in
            try OutboxEntry
                .filter(Column("state") == OutboxState.pending.rawValue)
                .filter(Column("nextAttemptAt") <= now)
                .order(Column("createdAt").asc)
                .fetchAll(db)
                .reduce(into: [String: OutboxEntry]()) { earliest, entry in
                    let key = "\(entry.aggregateType)|\(entry.aggregateId)"
                    if earliest[key] == nil { earliest[key] = entry }
                }
                .values
                .sorted { $0.createdAt < $1.createdAt }
                .prefix(limit)
                .map { $0 }
        }
    }

    public func markInFlight(_ entry: OutboxEntry) async throws {
        var updated = entry
        updated.state = .inFlight
        updated.lastAttemptAt = clock()
        try await database.write { db in try updated.update(db) }
    }

    public func markConfirmed(_ entry: OutboxEntry) async throws {
        _ = try await database.write { db in
            try OutboxEntry.deleteOne(db, key: entry.id)
        }
    }

    /// Records a terminal business rejection.
    ///
    /// The payload is deliberately retained. Discarding what someone typed because the server said
    /// no is hostile — the UI offers it back for editing (docs/sync-protocol.md §4.1).
    public func markRejected(_ entry: OutboxEntry, code: String?, message: String?) async throws {
        var updated = entry
        updated.state = .rejected
        updated.failureCode = code
        updated.failureMessage = message
        updated.lastAttemptAt = clock()
        try await database.write { db in try updated.update(db) }
    }

    /// Schedules a retry, or gives up if the entry has been trying for too long.
    public func scheduleRetry(_ entry: OutboxEntry) async throws {
        let now = clock()
        var updated = entry
        updated.attemptCount = entry.attemptCount + 1
        updated.lastAttemptAt = now

        if now.timeIntervalSince(entry.createdAt) > Self.retryDeadline {
            updated.state = .failed
            updated.failureCode = "RETRY_DEADLINE_EXCEEDED"
            updated.failureMessage = "Could not reach the server for 7 days"
        } else {
            updated.state = .pending
            updated.nextAttemptAt = now.addingTimeInterval(Self.backoff(attempt: updated.attemptCount))
        }

        try await database.write { db in try updated.update(db) }
    }

    /// Requeues entries stranded `inFlight` by a process termination. Safe: idempotency keys.
    @discardableResult
    public func recoverStranded() async throws -> Int {
        try await database.write { db in
            try OutboxEntry
                .filter(Column("state") == OutboxState.inFlight.rawValue)
                .updateAll(db, Column("state").set(to: OutboxState.pending.rawValue))
        }
    }

    /// Called on sign-out. Queued mutations are discarded after the user is warned.
    public func clear() async throws {
        _ = try await database.write { db in
            try OutboxEntry.deleteAll(db)
        }
    }

    public func pendingCount() async throws -> Int {
        try await database.read { db in
            try OutboxEntry
                .filter([OutboxState.pending.rawValue, OutboxState.inFlight.rawValue].contains(Column("state")))
                .fetchCount(db)
        }
    }

    /// Exponential backoff with **full jitter**.
    ///
    /// Jitter is not decoration: without it every device in a company retries in lockstep after an
    /// outage, and the recovering server is immediately knocked over again by its own users.
    /// Sampling uniformly from `[0, bound]` spreads the herd far better than a narrow band around
    /// the target.
    ///
    /// Kept identical to `Outbox.backoffMillis` on Android.
    static func backoff(attempt: Int) -> TimeInterval {
        let exponential = baseBackoff * pow(2, Double(min(attempt - 1, maxShift)))
        let capped = min(exponential, maxBackoff)
        return TimeInterval.random(in: 0...capped)
    }

    private static let baseBackoff: Double = 1
    private static let maxBackoff: Double = 5 * 60
    private static let maxShift = 20
    private static let retryDeadline: TimeInterval = 7 * 24 * 60 * 60
}

/// Sends one outbox entry. Implemented against the generated `HRClient`.
public protocol OutboxSender: Sendable {
    func send(_ entry: OutboxEntry) async -> SendOutcome
}

public enum SendOutcome: Sendable, Equatable {
    /// 2xx, or 409 `ALREADY_DECIDED` — someone else acted first, which is still a settled outcome.
    case confirmed
    /// Terminal 4xx. Will never succeed; the user's input is kept for editing.
    case rejected(code: String?, message: String?)
    /// 5xx or network. Retry with backoff.
    case retryable
    /// 401 that survived a token refresh. Pause the drain.
    case authenticationRequired
}

File: android/app/src/main/kotlin/com/hr/app/data/sync/Outbox.kt
package com.hr.app.data.sync

import com.hr.app.data.local.OutboxDao
import com.hr.app.data.local.OutboxEntry
import com.hr.app.data.local.OutboxState
import kotlinx.coroutines.flow.Flow
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.min
import kotlin.random.Random

/**
 * The write side of offline sync.
 *
 * Implements docs/sync-protocol.md §4. Callers enqueue a mutation and return immediately; the UI
 * has already been updated from the local write that accompanied it.
 */
@Singleton
class Outbox
    @Inject
    constructor(
        private val dao: OutboxDao,
        private val clock: Clock,
    ) {
        val pendingCount: Flow<Int> = dao.pendingCount()

        val problems: Flow<List<OutboxEntry>> = dao.problems()

        /**
         * Queues a mutation.
         *
         * The idempotency key is generated here, once. It must never be regenerated on retry —
         * that is precisely what makes unlimited retries safe, and regenerating it would turn a
         * lost response into a duplicate record.
         */
        suspend fun enqueue(
            aggregateType: String,
            aggregateId: String,
            httpMethod: String,
            path: String,
            payload: String,
        ): String {
            val id = UUID.randomUUID().toString()
            dao.insert(
                OutboxEntry(
                    id = id,
                    idempotencyKey = UUID.randomUUID().toString(),
                    aggregateType = aggregateType,
                    aggregateId = aggregateId,
                    httpMethod = httpMethod,
                    path = path,
                    payload = payload,
                    state = OutboxState.PENDING,
                    createdAt = clock.now(),
                ),
            )
            return id
        }

        suspend fun nextBatch(limit: Int = DRAIN_BATCH): List<OutboxEntry> = dao.nextBatch(clock.now(), limit)

        suspend fun markInFlight(entry: OutboxEntry) {
            dao.update(entry.copy(state = OutboxState.IN_FLIGHT, lastAttemptAt = clock.now()))
        }

        suspend fun markConfirmed(entry: OutboxEntry) {
            dao.delete(entry.id)
        }

        /**
         * Records a terminal business rejection.
         *
         * The payload is deliberately retained. Discarding what someone typed because the server
         * said no is hostile — the UI offers it back for editing (docs/sync-protocol.md §4.1).
         */
        suspend fun markRejected(
            entry: OutboxEntry,
            code: String?,
            message: String?,
        ) {
            dao.update(
                entry.copy(
                    state = OutboxState.REJECTED,
                    failureCode = code,
                    failureMessage = message,
                    lastAttemptAt = clock.now(),
                ),
            )
        }

        /**
         * Schedules a retry, or gives up if the entry has been trying for too long.
         *
         * Backoff is exponential **with jitter**. Jitter is not decoration: without it every
         * device in a company retries in lockstep after an outage, and the recovering server is
         * immediately knocked over again by its own users.
         */
        suspend fun scheduleRetry(entry: OutboxEntry) {
            val now = clock.now()
            val attempt = entry.attemptCount + 1

            if (now - entry.createdAt > RETRY_DEADLINE_MS) {
                dao.update(
                    entry.copy(
                        state = OutboxState.FAILED,
                        attemptCount = attempt,
                        lastAttemptAt = now,
                        failureCode = "RETRY_DEADLINE_EXCEEDED",
                        failureMessage = "Could not reach the server for 7 days",
                    ),
                )
                return
            }

            dao.update(
                entry.copy(
                    state = OutboxState.PENDING,
                    attemptCount = attempt,
                    lastAttemptAt = now,
                    nextAttemptAt = now + backoffMillis(attempt),
                ),
            )
        }

        /** Requeues entries stranded `IN_FLIGHT` by a process death. Safe: idempotency keys. */
        suspend fun recoverStranded(): Int = dao.requeueStranded()

        /** Called on sign-out. Queued mutations are discarded after the user is warned. */
        suspend fun clear() = dao.clear()

        internal fun backoffMillis(attempt: Int): Long {
            val exponential = BASE_BACKOFF_MS shl min(attempt - 1, MAX_SHIFT)
            val capped = min(exponential, MAX_BACKOFF_MS)
            // Full jitter: uniform in [0, capped]. Spreads a thundering herd far more effectively
            // than a small ± band around the target.
            return Random.nextLong(capped + 1)
        }

        private companion object {
            const val DRAIN_BATCH = 20
            const val BASE_BACKOFF_MS = 1_000L
            const val MAX_BACKOFF_MS = 5 * 60 * 1_000L
            const val MAX_SHIFT = 20
            const val RETRY_DEADLINE_MS = 7L * 24 * 60 * 60 * 1_000
        }
    }

/** Injected so tests can control time rather than sleeping. */
fun interface Clock {
    fun now(): Long
}
